runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -215,9 +218,11 @@ public static WatermarkUpdate fromTimestamps(Instant oldTime, Instant currentTim
     private final Collection<? extends Watermark> inputWatermarks;
     private final SortedMultiset<CommittedBundle<?>> pendingElements;
 
-    // This tracks only the quantity of timers at each timestamp, for quickly getting the cross-key
-    // minimum
-    private final SortedMultiset<Instant> pendingTimers;
+    // Keeps track of pending timers sorted by timestamp they fire at.
+    private final SortedMultiset<TimerData> pendingTimersSortedByTs;
+
+    // Keeps track of pending timers sorted by output timestamp.
+    private final SortedMultiset<TimerData> pendingTimersSortedByOutputTs;
```

Comment:
And for processing time timers, we just let them expire, because we don't want to have a pipeline just stall, right?